### PR TITLE
OH2-172 add apis to enhance opd manipulation with operations rows as it is done for oh 1 11

### PR DESCRIPTION
**Pull request by ArnaudFonzam**
See OH2-172.

Paired with https://github.com/informatici/openhospital-api/pull/215.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,5 +1,9 @@
+# Warning
+|:warning: This fork is for _prototyping purpose only_. If you want to submit a PR to the Open Hospital project starts from the [official ISF (Informatici Senza Frontiere) repos](https://github.com/informatici).      |
+|:---------------------------|
+
 # Open Hospital - Core
-[![Java CI](https://github.com/informatici/openhospital-core/workflows/Java%20CI%20with%20Maven/badge.svg)](https://github.com/informatici/openhospital-core/actions?query=workflow%3A%22Java+CI+with+Maven%22)
+[![Java CI](https://github.com/intesys/openhospital-core/workflows/Java%20CI%20with%20Maven/badge.svg)](https://github.com/intesys/openhospital-core/actions?query=workflow%3A%22Java+CI+with+Maven%22)
```
Comment (dbmalkovsky) [inline README.md:6]: please submit PRs against the Informatici repos and not local repos

File: pom.xml [hunk 2]
```diff
@@ -326,7 +326,7 @@
 		<dependency>
 			<groupId>org.igniterealtime.smack</groupId>
 			<artifactId>smackx</artifactId>
-			<version>3.4.1</version>
+			<version>3.2.1</version>
```
Comment (dbmalkovsky) [inline pom.xml:329]: do not downgrade jar versions

File: sql/step_71_permissions_and_grouppermission.sql [hunk 3]
```diff
@@ -0,0 +1,90 @@
+CREATE TABLE PERMISSIONS (
```
Comment (dbmalkovsky) [inline sql/step_71_permissions_and_grouppermission.sql:1]: isn't this step_86 that is already committed to the Core repo?

File: src/main/java/org/isf/menu/manager/UserGroupManager.java [hunk 4]
```diff
@@ -1,6 +1,10 @@
 /*
  * Open Hospital (www.open-hospital.org)
+<<<<<<< HEAD
+ * Copyright © 2006-2020 Informatici Senza Frontiere (info@informaticisenzafrontiere.org)
+=======
  * Copyright © 2006-2023 Informatici Senza Frontiere (info@informaticisenzafrontiere.org)
+>>>>>>> 2adacb561550c37e230459abb8801dad97229899
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/menu/manager/UserGroupManager.java:7]: do not downgrade copyright years

File: src/org/isf/patient/manager/PatientBrowserManager.java [hunk 5]
```diff
@@ -0,0 +1,402 @@
+package org.isf.patient.manager;
```
Comment (dbmalkovsky) [inline src/org/isf/patient/manager/PatientBrowserManager.java:1]: Missing license header

File: src/org/isf/patient/manager/PatientBrowserManager.java [hunk 6]
```diff
@@ -0,0 +1,402 @@
...
+
+import java.util.*;
```
Comment (dbmalkovsky) [inline src/org/isf/patient/manager/PatientBrowserManager.java:3]: The project standards do not allow wildcard imports; you need to enumerate the imprts.

File: src/org/isf/patient/manager/PatientBrowserManager.java [hunk 7]
```diff
@@ -0,0 +1,402 @@
...
+	/**
+	 * method that insert a new Patient in the db
+	 * 
+	 * @param patient
+	 * @return true - if the new Patient has been inserted
+	 * @throws OHServiceException 
+	 */
+	public boolean newPatient(Patient patient) throws OHServiceException {
+        validate(patient);
+        return ioOperations.newPatient(patient);
```
Comment (dbmalkovsky) [inline src/org/isf/patient/manager/PatientBrowserManager.java:53]: formatting is wrong

File: src/org/isf/patient/manager/PatientBrowserManager.java [hunk 8]
```diff
@@ -0,0 +1,402 @@
...
+	public PatientIoOperations getIoOperations() {
+		return ioOperations;
+	}
+
+	public void setIoOperations(PatientIoOperations ioOperations) {
+		this.ioOperations = ioOperations;
+	}
+	
+	/**
+	 * method that insert a new Patient in the db
```
Comment (dbmalkovsky) [inline src/org/isf/patient/manager/PatientBrowserManager.java:45]:
```suggestion
	 * Method that inserts a new Patient into the db
```

Comment (dbmalkovsky) [reply src/org/isf/patient/manager/PatientBrowserManager.java:45]: Javadoc comments start with a capital letter

File: src/org/isf/patient/manager/PatientBrowserManager.java [hunk 9]
```diff
@@ -0,0 +1,402 @@
...
+                    MessageBundle.getMessage("angal.admission.cannotmergeadmittedpatients"), OHSeverityLevel.ERROR));
+            errors.add(new OHExceptionMessage(MessageBundle.getMessage("angal.admission.merge"),
+                    MessageBundle.getMessage("angal.admission.patientscannothavependingtask"), OHSeverityLevel.INFO));
+        }
+
+        boolean billPending = false;
+        
+        ArrayList<Bill> bills = billManager.getPendingBills(mergedPatient.getCode());
+        bills = billManager.getPendingBills(mergedPatient.getCode());
+        if (bills != null && !bills.isEmpty()) billPending = true;
```
Comment (dbmalkovsky) [inline src/org/isf/patient/manager/PatientBrowserManager.java:236]: Missing braces on the `if`

File: src/org/isf/patient/manager/PatientBrowserManager.java [hunk 10]
```diff
@@ -0,0 +1,402 @@
...
+	/**
+	 * method that merge patients and all clinic details under the same PAT_ID
+	 * 
+	 * @param mergedPatient
+	 * @param patient2
+	 * @return true - if no OHServiceException occurred
+	 * @throws OHServiceException 
+	 */
+	public boolean mergePatient(Patient mergedPatient, Patient patient2) throws OHServiceException  {
+			if (mergedPatient.getBirthDate() != null &&
```
Comment (dbmalkovsky) [inline src/org/isf/patient/manager/PatientBrowserManager.java:296]: too much indentation?

File: src/org/isf/patient/manager/PatientBrowserManager.java [hunk 11]
```diff
@@ -0,0 +1,402 @@
...
+					mergedPatient.setBirthDate(bdate2);
+				}
+				if (bdate2 != null && ageType2.compareTo("") != 0) {
+					//patient2 has AgeType 
+					mergedPatient.setAge(age2);
+					mergedPatient.setAgetype(ageType2);
+				}
+			}
+			
+			if (mergedPatient.getAddress().compareTo("") == 0)
```
Comment (dbmalkovsky) [inline src/org/isf/patient/manager/PatientBrowserManager.java:314]: Missing braces here and on following lines

File: src/org/isf/patient/manager/PatientBrowserManager.java [hunk 12]
```diff
@@ -0,0 +1,402 @@
...
+			
+			if (mergedPatient.getNote().compareTo("") == 0)
+				mergedPatient.setNote(patient2.getNote());
+			else {
+				String note = mergedPatient.getNote();
+				mergedPatient.setNote(patient2.getNote()+"\n\n"+note);
+			}
+
+            List<OHExceptionMessage> errors = validateMergePatients(mergedPatient, patient2);
+            if(!errors.isEmpty()){
```
Comment (dbmalkovsky) [inline src/org/isf/patient/manager/PatientBrowserManager.java:355]:
```suggestion
            if (!errors.isEmpty()) {
```

File: src/org/isf/patient/manager/PatientBrowserManager.java [hunk 13]
```diff
@@ -0,0 +1,402 @@
...
+        }
+        if (!checkAge(patient)) {
+            errors.add(new OHExceptionMessage(MessageBundle.getMessage("angal.hospital"), MessageBundle.getMessage("angal.patient.insertvalidage"),
+                    OHSeverityLevel.ERROR));
+        }
+        if (StringUtils.isEmpty(String.valueOf(patient.getSex()))) {
+            errors.add(new OHExceptionMessage(MessageBundle.getMessage("angal.hospital"), "Please select a sex",
+                    OHSeverityLevel.ERROR));
+        }
+	    if(!errors.isEmpty()){
```
Comment (dbmalkovsky) [inline src/org/isf/patient/manager/PatientBrowserManager.java:385]: formatting

File: src/org/isf/patient/manager/PatientBrowserManager.java [hunk 14]
```diff
@@ -0,0 +1,402 @@
...
+	        throw new OHDataValidationException(errors);
+	    }
+    }
+
+    private boolean checkAge(Patient patient) {
+	    Date now = new Date();
```
Comment (dbmalkovsky) [inline src/org/isf/patient/manager/PatientBrowserManager.java:391]: formatting

File: README.md (hunk 1 above)
Comment (dbmalkovsky) [inline README.md:6]: Do not change repos; this must run against the Informatici repo and not your local one

**Comment by dbmalkovsky** [comment]
I stopped making comments because there are so many issues with this PR; it needs to be reviewed by the submitter and changes made

File: src/main/java/org/isf/menu/manager/UserGroupManager.java (hunk 4 above)
Comment (ArnaudFonzam) [reply src/main/java/org/isf/menu/manager/UserGroupManager.java:7]: okay

File: rsc/database.properties [hunk 15]
```diff
@@ -0,0 +1,3 @@
+jdbc.url=jdbc:mysql://localhost:3306/oh
```
[the remark this answers is not in the document]
Comment (ArnaudFonzam) [reply rsc/database.properties:1]: the .dist files are in the api not in the core. the core has only one file database.properties.

**Comment by ArnaudFonzam** [comment]
Hello @dbmalkovsky please i think all is okay know

File: rsc/database.properties (hunk 15 above)
[the remark this answers is not in the document]
Comment (dbmalkovsky) [reply rsc/database.properties:1]: The Core project doesn't need this file.   The `rsc` directory is only in the GUI repo.  For Core testing the database properties are found int `openhospital-core/src/test/resources/database.properties`

File: src/org/isf/patient/service/PatientIoOperationRepository.java [hunk 16]
```diff
@@ -0,0 +1,151 @@
+/*
+ * Open Hospital (www.open-hospital.org)
+ * Copyright © 2006-2021 Informatici Senza Frontiere (info@informaticisenzafrontiere.org)
```
Comment (dbmalkovsky) [inline src/org/isf/patient/service/PatientIoOperationRepository.java:3]: Change the end year to `2023`

File: src/org/isf/patient/service/PatientIoOperationRepository.java [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
+import org.springframework.stereotype.Repository;
+import org.springframework.transaction.annotation.Transactional;
+
+
+@Repository
+public interface PatientIoOperationRepository extends JpaRepository<Patient, Integer>, PatientIoOperationRepositoryCustom {
+    
+	@Query(value = "SELECT * FROM PATIENT WHERE (PAT_DELETED='N' OR PAT_DELETED IS NULL) ORDER BY PAT_ID", nativeQuery= true)
```
Comment (dbmalkovsky) [inline src/org/isf/patient/service/PatientIoOperationRepository.java:42]: formatting

File: src/org/isf/patient/service/PatientIoOperationRepositoryImpl.java [hunk 18]
```diff
@@ -0,0 +1,113 @@
+package org.isf.patient.service;
```
Comment (dbmalkovsky) [inline src/org/isf/patient/service/PatientIoOperationRepositoryImpl.java:1]: Missing license header.

File: src/org/isf/patient/service/PatientIoOperationRepositoryImpl.java [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+
+import javax.persistence.EntityManager;
+import javax.persistence.PersistenceContext;
+import javax.persistence.criteria.*;
```
Comment (dbmalkovsky) [inline src/org/isf/patient/service/PatientIoOperationRepositoryImpl.java:11]: No wildcard imports; enumerate them all.

**Comment by dbmalkovsky** [comment]
No it is far from OK; you never addressed any of the comments in `src/org/isf/patient/manager/PatientBrowserManager.java` and the same mistakes that I pointed in other files still exist in the ones that I didn't explicitly enumerate because I was hoping that you'd see the pattern of things that needed to be done and apply them to the other files.

File: rsc/database.properties (hunk 15 above)
[the remark this answers is not in the document]
Comment (mwithi) [reply rsc/database.properties:1]: Correct, the "core" project only has a `.example` file, the only thing the core needs is to find .properties in the classpath when "gui" or "api" are launched using "core" as dependency (pretty much like it happens with log4j)

[the remark this answers is not in the document]
Comment (ArnaudFonzam) [reply rsc/database.properties:1]: okay

File: src/org/isf/patient/service/PatientIoOperationRepository.java (hunk 17 above)
Comment (mwithi) [reply src/org/isf/patient/service/PatientIoOperationRepository.java]: how this was resolved? I can see the indentation is still wrong, also in other lines of this file

File: src/org/isf/patient/service/PatientIoOperationRepositoryImpl.java (hunk 18 above)
Comment (mwithi) [reply src/org/isf/patient/service/PatientIoOperationRepositoryImpl.java]: This review is not resolved, @ArnaudFonzam you need to add the License header to all new files

File: src/org/isf/patient/service/PatientIoOperationRepositoryImpl.java (hunk 19 above)
Comment (mwithi) [reply src/org/isf/patient/service/PatientIoOperationRepositoryImpl.java]: This is not resolved, @ArnaudFonzam you need to remove `import javax.persistence.criteria.*;` (that uses wildcards) and replace it with punctual import. You can use your IDE for that (in Eclipse -> "organize import" function)

File: src/org/isf/patient/service/PatientIoOperations.java [hunk 20]
```diff
@@ -0,0 +1,423 @@
+package org.isf.patient.service;
```
Comment (mwithi) [inline src/org/isf/patient/service/PatientIoOperations.java]: License header

File: src/org/isf/patient/test/Tests.java [hunk 21]
```diff
@@ -0,0 +1,429 @@
...
+
+    @BeforeClass
+    public static void setUpClass()  
+    {
+    	jpa = new DbJpaUtil();
+    	testPatient = new TestPatient();
+    	testPatientContext = new TestPatientContext();
+
+        return;
+    }
```
Comment (mwithi) [inline src/org/isf/patient/test/Tests.java]: All this file is wrongly formatted, @ArnaudFonzam please use https://github.com/informatici/openhospital-core/tree/develop/.ide-settings

File: README.md [hunk 22]
```diff
@@ -1,3 +1,7 @@
+# Warning
+|:warning: This fork is for _prototyping purpose only_. If you want to submit a PR to the Open Hospital project starts from the [official ISF (Informatici Senza Frontiere) repos](https://github.com/informatici).      |
+|:---------------------------|
+
```
Comment (mwithi) [inline README.md:4]: this part comes from Intesys but makes no sense in informatici repository, please remove it

File: src/org/isf/patient/service/PatientIoOperationRepository.java (hunk 16 above)
Comment (mwithi) [reply src/org/isf/patient/service/PatientIoOperationRepository.java]: ok, the whole file has been deleted

File: src/org/isf/patient/service/PatientIoOperationRepository.java (hunk 17 above)
Comment (mwithi) [reply src/org/isf/patient/service/PatientIoOperationRepository.java]: ok, the whole file has been deleted

File: src/org/isf/patient/service/PatientIoOperationRepositoryImpl.java (hunk 18 above)
Comment (mwithi) [reply src/org/isf/patient/service/PatientIoOperationRepositoryImpl.java]: ok, the whole file has been deleted

File: src/org/isf/patient/service/PatientIoOperationRepositoryImpl.java (hunk 19 above)
Comment (mwithi) [reply src/org/isf/patient/service/PatientIoOperationRepositoryImpl.java]: ok, the whole file has been deleted

File: src/org/isf/patient/service/PatientIoOperations.java (hunk 20 above)
Comment (mwithi) [reply src/org/isf/patient/service/PatientIoOperations.java]: ok, the whole file has been deleted

File: src/org/isf/patient/test/Tests.java [hunk 23]
```diff
@@ -0,0 +1,429 @@
+package org.isf.patient.test;
```
[the remark this answers is not in the document]
Comment (mwithi) [reply src/org/isf/patient/test/Tests.java]: ok, the whole file has been deleted

File: src/org/isf/patient/test/Tests.java (hunk 21 above)
Comment (mwithi) [reply src/org/isf/patient/test/Tests.java]: ok, the whole file has been deleted

File: src/main/java/org/isf/patient/manager/PatientBrowserManager.java [hunk 24]
```diff
@@ -59,6 +59,14 @@ public class PatientBrowserManager {
...
 	protected LinkedHashMap<String, String> professionHashMap;
 
+	public PatientIoOperations getIoOperations() {
+		return ioOperations;
+	}
+
+	public void setIoOperations(PatientIoOperations ioOperations) {
+		this.ioOperations = ioOperations;
+	}
+
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/patient/manager/PatientBrowserManager.java:69]: Why are these two methods here?   The field `ioOPerations` is `@Autowired`; there is no need to set it.

Comment (mwithi) [reply src/main/java/org/isf/patient/manager/PatientBrowserManager.java:69]: yeah, also they seems not used in the whole stack

File: src/test/java/org/isf/operation/test/Tests.java [hunk 25]
```diff
@@ -816,7 +816,7 @@ public void testMgrRowUpdateOperationRowNotFound() throws Exception {
 		Operation operation = testOperation.setup(operationType, true);
 		OperationRow operationRow = testOperationRow.setup(operation, true);
 
-		assertThat(operationRowBrowserManager.updateOperationRow(operationRow)).isTrue();
+		assertThat(operationRowBrowserManager.updateOperationRow(operationRow));
```
Comment (dbmalkovsky) [inline src/test/java/org/isf/operation/test/Tests.java:819]: This `assertThat` isn't testing anything.

File: src/test/java/org/isf/operation/test/Tests.java [hunk 26]
```diff
@@ -825,7 +825,7 @@ public void testMgrRowUpdateOperationRow() throws Exception {
 		OperationRow operationRow = operationRowIoOperationRepository.findById(id);
 		assertThat(operationRow.getRemarks()).isNotEqualTo("someNewRemarks");
 		operationRow.setRemarks("someNewRemarks");
-		assertThat(operationRowBrowserManager.updateOperationRow(operationRow)).isTrue();
+		assertThat(operationRowBrowserManager.updateOperationRow(operationRow));
```
Comment (dbmalkovsky) [inline src/test/java/org/isf/operation/test/Tests.java:828]: OperationRow operationRow = operationRowIoOperationRepository.findById(id);

File: src/test/java/org/isf/operation/test/Tests.java [hunk 27]
```diff
@@ -839,7 +839,7 @@ public void testMgrRowNewOperationRow() throws Exception {
 		operationTypeIoOperationRepository.saveAndFlush(operationType);
 		operationIoOperationRepository.saveAndFlush(operation);
 
-		assertThat(operationRowBrowserManager.newOperationRow(operationRow)).isTrue();
+		assertThat(operationRowBrowserManager.newOperationRow(operationRow));
```
Comment (dbmalkovsky) [inline src/test/java/org/isf/operation/test/Tests.java:842]: OperationRow operationRow = operationRowIoOperationRepository.findById(id);

File: src/main/java/org/isf/patient/manager/PatientBrowserManager.java (hunk 24 above)
Comment (ArnaudFonzam) [reply src/main/java/org/isf/patient/manager/PatientBrowserManager.java:69]: okay i will remove its methods.

**Comment by mwithi** [comment]
Hello @ArnaudFonzam, can you please fix test as done here in `src/test/java/org/isf/operation/test/Tests.java`? Thanks

https://github.com/informatici/openhospital-core/pull/886/files#diff-c9832a89d0489f8b27825f092ae45a5f365b91a93c7768f388110047c7142df2

**Review by mwithi — APPROVED** [review]
LGTM. Thanks a lot!

**Pull request merged**